File: src/Circle.AI.Core/ModelDownloader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;

namespace Circle.AI.Core
{
    /// <summary>
    /// Source-agnostic model downloader. Walks a list of <see cref="IModelSource"/>
    /// instances in order, falling through on failure so that one supplier going dark
    /// (sanctions, regional blocks, outages) does not break model bootstrap.
    /// </summary>
    public sealed class ModelDownloader : IModelDownloader, IDisposable
    {
        /// <summary>
        /// Progress report shape emitted during downloads.
        /// Mirrors <see cref="DownloadProgress"/> as a class+event for consumer compatibility.
        /// </summary>
        public sealed class DownloadProgressReport
        {
            public string FileName { get; set; } = "";
            public long BytesReceived { get; set; }
            public long TotalBytes { get; set; }
            public double BytesPerSecond { get; set; }
            public TimeSpan EstimatedTimeRemaining { get; set; }
        }

        public delegate void DownloadProgressHandler(DownloadProgressReport progress);
        public event DownloadProgressHandler? ProgressChanged;

        private const string RegistryResourceName = "Circle.AI.Core.registry.json";

        private readonly IReadOnlyList<IModelSource> _sources;
        private readonly bool _ownsSources;
        private readonly Lazy<IReadOnlyDictionary<string, ModelEntry>> _registry;
        private bool _disposed;

        public ModelDownloader(IReadOnlyList<IModelSource> sources, bool ownsSources = false)
        {
            if (sources is null) throw new ArgumentNullException(nameof(sources));
            if (sources.Count == 0)
                throw new ArgumentException("At least one model source is required", nameof(sources));

            _sources = sources;
            _ownsSources = ownsSources;
            _registry = new Lazy<IReadOnlyDictionary<string, ModelEntry>>(LoadEmbeddedRegistry);
        }

        /// <inheritdoc />
        public async Task DownloadModelAsync(string modelId, string localPath, CancellationToken ct = default)
        {
            if (_disposed) throw new ObjectDisposedException(nameof(ModelDownloader));
            if (string.IsNullOrWhiteSpace(modelId)) throw new ArgumentNullException(nameof(modelId));
            if (string.IsNullOrWhiteSpace(localPath)) throw new ArgumentNullException(nameof(localPath));

            if (!_registry.Value.TryGetValue(modelId, out var entry))
            {
                throw new KeyNotFoundException(
                    $"Model '{modelId}' is not in the embedded registry. Known models: " +
                    string.Join(", ", _registry.Value.Keys));
            }

            Directory.CreateDirectory(localPath);
            var targetFile = Path.Combine(localPath, entry.FileName);

            var candidates = BuildCandidateList(entry);
            if (candidates.Count == 0)
            {
                throw new InvalidOperationException(
                    $"Model '{modelId}' has no PrimaryUrl or FallbackUrl configured.");
            }

            var bridge = new Progress<DownloadProgress>(p =>
                ProgressChanged?.Invoke(new DownloadProgressReport
                {
                    FileName = p.FileName,
                    BytesReceived = p.BytesReceived,
                    TotalBytes = p.TotalBytes,
                    BytesPerSecond = p.BytesPerSecond,
                    EstimatedTimeRemaining = p.EstimatedTimeRemaining,
                }));

            try
            {
                var winner = await DownloadFromCandidatesAsync(candidates, targetFile, bridge, ct).ConfigureAwait(false);
                Console.WriteLine($"[ModelDownloader] '{modelId}' downloaded via {winner}.");
            }
            catch
            {
                CleanupPartialFile(targetFile);
                throw;
            }
        }

        /// <inheritdoc />
        public async Task<string> DownloadFromCandidatesAsync(
            IReadOnlyList<string> candidateUrls,
            string localFilePath,
            IProgress<DownloadProgress>? progress = null,
            CancellationToken ct = default)
        {
            if (_disposed) throw new ObjectDisposedException(nameof(ModelDownloader));
            if (candidateUrls is null) throw new ArgumentNullException(nameof(candidateUrls));
            if (candidateUrls.Count == 0)
                throw new ArgumentException("At least one candidate URL is required", nameof(candidateUrls));
            if (string.IsNullOrWhiteSpace(localFilePath)) throw new ArgumentNullException(nameof(localFilePath));

            var dir = Path.GetDirectoryName(localFilePath);
            if (!string.IsNullOrEmpty(dir)) Directory.CreateDirectory(dir);

            var failures = new List<string>();

            foreach (var url in candidateUrls)
            {
                ct.ThrowIfCancellationRequested();
                if (string.IsNullOrWhiteSpace(url)) continue;

                var source = MatchSource(url);
                if (source is null)
                {
                    Console.WriteLine($"[ModelDownloader] Warning: no registered source matched URL '{url}' — skipping. Add a source whose Name matches the hostname, or extend MatchSource.");
                    failures.Add($"(no registered source for '{url}')");
                    continue;
                }

                try
                {
                    Console.WriteLine($"[ModelDownloader] Trying {source.Name}: {url}");
                    await source.DownloadAsync(url, localFilePath, progress, ct).ConfigureAwait(false);
                    Console.WriteLine($"[ModelDownloader] {source.Name} succeeded.");
                    return source.Name;
                }
                catch (OperationCanceledException) when (ct.IsCancellationRequested)
                {
                    throw;
                }
                catch (Exception ex)
                {
                    failures.Add($"{source.Name}: {ex.Message}");
                    Console.WriteLine($"[ModelDownloader] {source.Name} failed: {ex.Message}. Falling through.");
                    // Drop the partial so the next source can start clean.
                    CleanupPartialFile(localFilePath);
                }
            }

            throw new InvalidOperationException(
                "All model sources failed:\n  " + string.Join("\n  ", failures));
        }

        private IModelSource? MatchSource(string url)
        {
            if (!Uri.TryCreate(url, UriKind.Absolute, out var uri)) return null;
            var host = uri.Host;

            // Heuristic match by source Name, then by host substring.
            foreach (var s in _sources)
            {
                if (host.Contains(s.Name, StringComparison.OrdinalIgnoreCase))
                    return s;
            }

            // All registered download URLs are on modelscope.cn (Alibaba).
            // No Western source is registered — if a URL slips through that isn't
            // on a registered source, MatchSource returns null and the downloader skips it.
            if (host.Contains("modelscope", StringComparison.OrdinalIgnoreCase))
                return _sources.FirstOrDefault(s => s.Name.Equals("ModelScope", StringComparison.OrdinalIgnoreCase));

            return null;
        }

        private static IReadOnlyList<string> BuildCandidateList(ModelEntry entry)
        {
            var list = new List<string>(2);
            if (!string.IsNullOrWhiteSpace(entry.PrimaryUrl)) list.Add(entry.PrimaryUrl);
            if (!string.IsNullOrWhiteSpace(entry.FallbackUrl)) list.Add(entry.FallbackUrl);
            return list;
        }

        private static IReadOnlyDictionary<string, ModelEntry> LoadEmbeddedRegistry()
        {
            var assembly = typeof(ModelDownloader).Assembly;
            using var stream = assembly.GetManifestResourceStream(RegistryResourceName);
            if (stream is null)
            {
                // Registry isn't embedded — fall back to a sibling registry.json next to the assembly.
                var assemblyDir = Path.GetDirectoryName(assembly.Location);
                var fallback = assemblyDir is null
                    ? null
                    : Path.Combine(assemblyDir, "registry.json");

                if (fallback is null || !File.Exists(fallback))
                {
                    return new Dictionary<string, ModelEntry>(StringComparer.OrdinalIgnoreCase);
                }

                using var fs = File.OpenRead(fallback);
                return JsonSerializer.Deserialize<Dictionary<string, ModelEntry>>(fs, JsonOpts())
                       ?? new Dictionary<string, ModelEntry>(StringComparer.OrdinalIgnoreCase);
            }

            return JsonSerializer.Deserialize<Dictionary<string, ModelEntry>>(stream, JsonOpts())
                   ?? new Dictionary<string, ModelEntry>(StringComparer.OrdinalIgnoreCase);
        }

        private static JsonSerializerOptions JsonOpts() => new()
        {
            PropertyNameCaseInsensitive = true,
            ReadCommentHandling = JsonCommentHandling.Skip,
            AllowTrailingCommas = true,
        };

        private static void CleanupPartialFile(string path)
        {
            try
            {
                if (File.Exists(path)) File.Delete(path);
            }
            catch
            {
                // Best effort.
            }
        }

        public void Dispose()
        {
            if (_disposed) return;
            if (_ownsSources)
            {
                foreach (var s in _sources)
                {
                    (s as IDisposable)?.Dispose();
                }
            }
            _disposed = true;
        }

        private sealed record ModelEntry
        {
            [JsonPropertyName("FileName")] public string FileName { get; init; } = "";
            [JsonPropertyName("PrimaryUrl")] public string? PrimaryUrl { get; init; }
            [JsonPropertyName("FallbackUrl")] public string? FallbackUrl { get; init; }
            [JsonPropertyName("Checksum")] public string? Checksum { get; init; }
            [JsonPropertyName("SizeBytes")] public long SizeBytes { get; init; }
            [JsonPropertyName("Version")] public string? Version { get; init; }
            [JsonPropertyName("Architecture")] public string? Architecture { get; init; }
            [JsonPropertyName("QuantizationType")] public string? QuantizationType { get; init; }
        }
    }
}

File: go/model_downloader.go
// model_downloader.go
//
// Ports CircleAI.Core.IModelDownloader (IModelDownloader.cs) and
// CircleAI.Core.ModelDownloader (+ nested DownloadProgressReport)
// (ModelDownloader.cs).
//
// Source-agnostic downloader: walks a list of IModelSource in order, falling
// through on failure so one supplier going dark does not break bootstrap.
//
// Divergence from C#, intentional and behaviour-preserving: C# reads the model
// set from an assembly-embedded registry.json. Go has no embedded-resource
// concept, so the registry is injected as a map[string]DownloaderModelEntry via
// the constructor (the flat package already owns ModelEntry/ModelRegistry for
// the richer catalog path). All routing, fallback, bundle-rejection, and the
// progress-bridge logic match the C# reference exactly.

package circleai

import (
	"context"
	"errors"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// DownloadProgressReport mirrors ModelDownloader.DownloadProgressReport — the
// class+callback progress shape emitted to consumers of the high-level
// DownloadModel entry point.
type DownloadProgressReport struct {
	FileName               string
	BytesReceived          int64
	TotalBytes             int64
	BytesPerSecond         float64
	EstimatedTimeRemaining time.Duration
}

// DownloaderModelEntry is the registry row the downloader resolves a modelId
// to. Mirrors ModelDownloader.ModelEntry (single-file legacy shape + bundle
// shape). Named DownloaderModelEntry to avoid colliding with the catalog-level
// ModelEntry already in registry.go.
type DownloaderModelEntry struct {
	// Legacy single-file shape.
	FileName         string
	PrimaryURL       string
	FallbackURL      string
	Checksum         string
	SizeBytes        int64
	Version          string
	Architecture     string
	QuantizationType string

	// Bundle shape.
	Repo        string
	TotalBytes  int64
	BundleFiles []BundleFile
}

// IsBundle reports whether this entry is a multi-file bundle.
func (e DownloaderModelEntry) IsBundle() bool { return len(e.BundleFiles) > 0 }

// IModelDownloader downloads a model (file or set) to local storage.
// Ports CircleAI.Core.IModelDownloader.
type IModelDownloader interface {
	// DownloadModel downloads the model identified by modelId to localPath,
	// resolving the URL set internally.
	DownloadModel(ctx context.Context, modelId, localPath string) error

	// DownloadFromCandidates downloads a single file by trying each candidate
	// URL in order (first = primary, rest = fallbacks). Returns the name of the
	// source that succeeded.
	DownloadFromCandidates(
		ctx context.Context,
		candidateURLs []string,
		localFilePath string,
		progress func(SourceDownloadProgress),
	) (string, error)
}

// ModelDownloader is the source-agnostic IModelDownloader implementation.
// Ports CircleAI.Core.ModelDownloader.
type ModelDownloader struct {
	sources  []IModelSource
	registry map[string]DownloaderModelEntry

	// ProgressChanged, when non-nil, receives high-level progress reports from
	// DownloadModel (mirrors the C# ProgressChanged event).
	ProgressChanged func(DownloadProgressReport)
}

// NewModelDownloader builds a downloader over the given sources and registry.
// At least one source is required (mirrors the C# ctor guard). The registry map
// may be nil/empty; DownloadModel then reports the model as unknown.
func NewModelDownloader(sources []IModelSource, registry map[string]DownloaderModelEntry) (*ModelDownloader, error) {
	if len(sources) == 0 {
		return nil, errors.New("at least one model source is required")
	}
	reg := make(map[string]DownloaderModelEntry, len(registry))
	for k, v := range registry {
		reg[k] = v
	}
	return &ModelDownloader{sources: sources, registry: reg}, nil
}

// DownloadModel resolves modelId in the registry and downloads its file,
// trying PrimaryUrl then FallbackUrl. Bundle entries are rejected with a
// steer-to-the-right-path error, exactly like the C# reference.
func (d *ModelDownloader) DownloadModel(ctx context.Context, modelId, localPath string) error {
	if strings.TrimSpace(modelId) == "" {
		return errors.New("modelId is required")
	}
	if strings.TrimSpace(localPath) == "" {
		return errors.New("localPath is required")
	}

	entry, ok := d.registry[modelId]
	if !ok {
		return fmt.Errorf("model %q is not in the registry. Known models: %s",
			modelId, strings.Join(d.knownModelIds(), ", "))
	}

	if err := os.MkdirAll(localPath, 0o755); err != nil {
		return err
	}

	if entry.IsBundle() {
		return fmt.Errorf(
			"model %q is a multi-file bundle (registry entry has BundleFiles[]); "+
				"use the multi-file bundle downloader instead — this legacy single-file "+
				"downloader cannot fetch a multi-file bundle", modelId)
	}

	targetFile := filepath.Join(localPath, entry.FileName)
	candidates := buildCandidateList(entry)
	if len(candidates) == 0 {
		return fmt.Errorf("model %q has no PrimaryUrl or FallbackUrl configured", modelId)
	}

	bridge := func(p SourceDownloadProgress) {
		if d.ProgressChanged != nil {
			d.ProgressChanged(DownloadProgressReport{
				FileName:               p.FileName,
				BytesReceived:          p.BytesReceived,
				TotalBytes:             p.TotalBytes,
				BytesPerSecond:         p.BytesPerSecond,
				EstimatedTimeRemaining: p.EstimatedTimeRemaining,
			})
		}
	}

	if _, err := d.DownloadFromCandidates(ctx, candidates, targetFile, bridge); err != nil {
		cleanupPartialFile(targetFile)
		return err
	}
	return nil
}

// DownloadFromCandidates tries each candidate URL against the first source
// whose Name/host matches, falling through on failure and cleaning up the
// partial file between attempts. Ports DownloadFromCandidatesAsync.
func (d *ModelDownloader) DownloadFromCandidates(
	ctx context.Context,
	candidateURLs []string,
	localFilePath string,
	progress func(SourceDownloadProgress),
) (string, error) {
	if len(candidateURLs) == 0 {
		return "", errors.New("at least one candidate URL is required")
	}
	if strings.TrimSpace(localFilePath) == "" {
		return "", errors.New("localFilePath is required")
	}
	if dir := filepath.Dir(localFilePath); dir != "" {
		if err := os.MkdirAll(dir, 0o755); err != nil {
			return "", err
		}
	}

	var failures []string
	for _, u := range candidateURLs {
		if err := ctx.Err(); err != nil {
			return "", err
		}
		if strings.TrimSpace(u) == "" {
			continue
		}
		src := d.matchSource(u)
		if src == nil {
			failures = append(failures, fmt.Sprintf("(no registered source for %q)", u))
			continue
		}
		if err := src.Download(ctx, u, localFilePath, progress); err != nil {
			if ctx.Err() != nil {
				return "", ctx.Err()
			}
			failures = append(failures, fmt.Sprintf("%s: %v", src.Name(), err))
			cleanupPartialFile(localFilePath)
			continue
		}
		return src.Name(), nil
	}
	return "", fmt.Errorf("all model sources failed:\n  %s", strings.Join(failures, "\n  "))
}

// matchSource picks the source whose Name appears in the URL host, then falls
// back to a modelscope substring rule. Ports ModelDownloader.MatchSource.
func (d *ModelDownloader) matchSource(rawURL string) IModelSource {
	u, err := url.Parse(rawURL)
	if err != nil || u.Host == "" {
		return nil
	}
	host := u.Host
	for _, s := range d.sources {
		if strings.Contains(strings.ToLower(host), strings.ToLower(s.Name())) {
			return s
		}
	}
	if strings.Contains(strings.ToLower(host), "modelscope") {
		for _, s := range d.sources {
			if strings.EqualFold(s.Name(), "ModelScope") {
				return s
			}
		}
	}
	return nil
}

func (d *ModelDownloader) knownModelIds() []string {
	ids := make([]string, 0, len(d.registry))
	for k := range d.registry {
		ids = append(ids, k)
	}
	sort.Strings(ids)
	return ids
}

func buildCandidateList(e DownloaderModelEntry) []string {
	list := make([]string, 0, 2)
	if strings.TrimSpace(e.PrimaryURL) != "" {
		list = append(list, e.PrimaryURL)
	}
	if strings.TrimSpace(e.FallbackURL) != "" {
		list = append(list, e.FallbackURL)
	}
	return list
}

func cleanupPartialFile(path string) {
	if _, err := os.Stat(path); err == nil {
		_ = os.Remove(path)
	}
}

var _ IModelDownloader = (*ModelDownloader)(nil)
